### PR TITLE
[PATCH API-NEXT v5] example: ipsec_offload: add ipsec_offload example

### PR DESCRIPTION
**Pull request by lumag**
This is based on old IPsec example by @NikhilA-Linaro , just updated to new packet-API.

[DES: adapted example to updated API.]

Signed-off-by: Nikhil Agarwal <nikhil.agarwal@linaro.org>
Signed-off-by: Dmitry Eremin-Solenikov <dmitry.ereminsolenikov@linaro.org>

### CODE REVIEW

**Comment by muvarov** [comment]
for v2: this can be fixed:
WARNING: else is not generally useful after a break or return
#435: FILE: example/ipsec_offload/odp_ipsec_offload.c:343:
+		return PKT_POSTED;
+	} else {

**Comment by lumag** [comment]
Current code is better, so I don't plan to fix it

File: example/ipsec_offload/odp_ipsec_offload.c [hunk 1]
```diff
@@ -0,0 +1,855 @@
...
+			odp_route_flow_insert_in_bucket(flow, &flow_table[hash &
+							(bucket_count - 1)]);
+			goto do_opt;
+		} else {
+			EXAMPLE_DBG("No flow match found. Packet is dropped.\n");
+			odp_packet_free(pkt);
+			return PKT_DROP;
+		}
+	}
+}
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_offload/odp_ipsec_offload.c]: Checkpatch doesn't like this and the `goto` is a bit ugly. Might a restructure along the following lines be simpler and cleaner?
```
if (!flow) {
        /*Check into Routing table*/
        fwd_entry = find_fwd_db_entry(dip);
     
        if (!fwd_entry) {
			EXAMPLE_DBG("No flow match found. Packet is dropped.\n");
			odp_packet_free(pkt);
 	                return PKT_DROP;
        }
        ... flow setup processing, fall though without needing a goto.
}

...flow processing goes here
```

File: example/ipsec_offload/odp_ipsec_offload_fwd_db.c [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * bucket count. It will be updated with user argument if provided
+ */
+uint32_t bucket_count = DEFAULT_BUCKET_COUNT;
+
+/** Global pointer to fwd db */
+fwd_db_t *fwd_db;
+
+void odp_init_routing_table(void)
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_offload/odp_ipsec_offload_fwd_db.c]: Avoid `odp_` prefix use for non-APIs.

File: example/ipsec_offload/odp_ipsec_offload_fwd_db.h [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+ * @param sip           Source IP Address
+ * @param dip           Destination IP Address
+ * @param sport         Source Port Number
+ * @param dport         Destination Port Number
+ * @param proto         IP protocol
+ * @param bucket        Hash Bucket
+ *
+ * @return Matching flow entry
+ */
+static inline odp_flow_entry_t *odp_route_flow_lookup_in_bucket(uint32_t sip,
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_offload/odp_ipsec_offload_fwd_db.h]: Drop `odp_` prefix

File: example/ipsec_offload/odp_ipsec_offload_fwd_db.h [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+/**
+ * Insert the flow into given hash bucket
+ *
+ * @param flow		Which is to be inserted
+ * @param bucket	Target Hash Bucket
+ *
+ */
+static inline void odp_route_flow_insert_in_bucket(odp_flow_entry_t *flow,
```
Comment (Bill-Fischofer-Linaro) [inline example/ipsec_offload/odp_ipsec_offload_fwd_db.h]: Drop `odp_` prefix.

**Comment by lumag** [comment]
@Bill-Fischofer-Linaro I have dropped `odp_` prefix from function names and reworked this `if (flow)` part. Regarding files names. All examples use `odp_` prefix for file names, so I'm not changing that.

 @bala-manoharan @NikhilA-Linaro @psavol I feel uneasy about using 'both' to specify ESP packet with  authentication and just 'esp' to specify encrypted ESP packet w/o auth. I might want to change that in next revision.

File: example/ipsec_offload/Makefile.am [hunk 5]
```diff
@@ -0,0 +1,19 @@
+include $(top_srcdir)/example/Makefile.inc
+
+bin_PROGRAMS = odp_ipsec_offload$(EXEEXT)
```
Comment (muvarov) [inline example/ipsec_offload/Makefile.am]: EXEEXT needs to be removed.

File: example/ipsec_offload/Makefile.am [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_misc.h \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_sa_db.h \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_sp_db.h \
+		  $(top_srcdir)/example/example_debug.h
+
+dist_odp_ipsec_offload_SOURCES = odp_ipsec_offload.c \
+			 odp_ipsec_offload_sa_db.c \
+			 odp_ipsec_offload_sp_db.c \
+			 odp_ipsec_offload_fwd_db.c \
+			 odp_ipsec_offload_cache.c
```
Comment (muvarov) [inline example/ipsec_offload/Makefile.am]: EXTRA_DIST for bash scripts which should be named correctly.

File: example/ipsec_offload/Makefile.am [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+noinst_HEADERS = \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_cache.h \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_fwd_db.h \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_misc.h \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_sa_db.h \
+		  $(top_srcdir)/example/ipsec_offload/odp_ipsec_offload_sp_db.h \
+		  $(top_srcdir)/example/example_debug.h
+
+dist_odp_ipsec_offload_SOURCES = odp_ipsec_offload.c \
```
Comment (muvarov) [inline example/ipsec_offload/Makefile.am]: just _SOURCES like all other example/*/Makefile.am

File: example/ipsec_offload/Makefile.am [hunk 8]
```diff
@@ -0,0 +1,19 @@
+include $(top_srcdir)/example/Makefile.inc
+
+bin_PROGRAMS = odp_ipsec_offload$(EXEEXT)
+odp_ipsec_offload_LDFLAGS = $(AM_LDFLAGS) -static
+odp_ipsec_offload_CFLAGS = $(AM_CFLAGS) -I${top_srcdir}/example
```
Comment (muvarov) [inline example/ipsec_offload/Makefile.am]: no need for CFLAGS/LDFLAGS

File: example/ipsec_offload/odp_ipsec_offload_cache.c [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+#define IPDEFTTL 64
+
+void init_ipsec_cache(void)
+{
+	odp_shm_t shm;
+
+	shm = odp_shm_reserve("shm_ipsec_cache",
+			      sizeof(ipsec_cache_t),
+			      ODP_CACHE_LINE_SIZE,
+			      0);
```
Comment (muvarov) [inline example/ipsec_offload/odp_ipsec_offload_cache.c]: shm != ODP_SHM_INVALID

File: example/ipsec_offload/odp_ipsec_offload_fwd_db.c [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+void init_routing_table(void)
+{
+	odp_shm_t		hash_shm;
+	uint32_t		i;
+	flow_bucket_t		*bucket;
+
+	/*Reserve memory for Routing hash table*/
+	hash_shm = odp_shm_reserve("route_table",
+				   sizeof(flow_bucket_t) * bucket_count,
+				   ODP_CACHE_LINE_SIZE, 0);
```
Comment (muvarov) [inline example/ipsec_offload/odp_ipsec_offload_fwd_db.c]: != ODP_SHM_INVALID

File: example/ipsec_offload/odp_ipsec_offload_fwd_db.h [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+ *
+ * @param input  Pointer to string describing route
+ *
+ * @param if_names  Array of Name of the interfaces available
+ *
+ * @param if_count  number of interfaces in if_names array
+ *
+ * @param entries number of entries
+ *
+ * @return 0 if successful else -1
```
Comment (muvarov) [inline example/ipsec_offload/odp_ipsec_offload_fwd_db.h]: why this file has bunch of empty lines like here? On  other files code is more compact. Let's keep the same code style.

File: example/ipsec_offload/odp_ipsec_offload_misc.h [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+extern "C" {
+#endif
+
+#include <odp.h>
+#include <odp/helper/eth.h>
+#include <odp/helper/ip.h>
+#include <odp/helper/ipsec.h>
+
+#ifndef TRUE
+#define TRUE  1
```
Comment (muvarov) [inline example/ipsec_offload/odp_ipsec_offload_misc.h]: looks a little bit ugly, but it's ok.

**Comment by muvarov** [comment]
@lumag git commit has changelog inside it. This needs to be cleaned up.

**Comment by lumag** [comment]
@muvarov updated

**Comment by lumag** [comment]
@muvarov It's not quite a changelog. It is a typical notion of changes done by person who applied the commit, rather than by original author.

File: example/ipsec_offload/odp_ipsec_offload_misc.h (hunk 12 above)
Comment (lumag) [reply example/ipsec_offload/odp_ipsec_offload_misc.h]: I did not want to be too intrusive into this example.

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**